// resin-multibuild/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Promise = require("bluebird");
const _ = require("lodash");
const tar = require("tar-stream");
function drainStream(stream) {
    return new Promise((resolve, reject) => {
        stream.on('data', _.noop);
        stream.on('error', reject);
        stream.on('end', resolve);
    });
}
exports.drainStream = drainStream;
function generateBuildTasks(images) {
    return _.map(images, (img) => {
        if (_.isString(img.image)) {
            return {
                external: true,
                imageName: img.image,
                serviceName: img.serviceName,
                resolved: false,
            };
        }
        else {
            return _.merge({
                external: false,
                serviceName: img.serviceName,
                buildStream: tar.pack(),
                resolved: false,
            }, img.image);
        }
    });
}
exports.generateBuildTasks = generateBuildTasks;
