// balena-cli/build/actions/internal.js
// Generated by CoffeeScript 1.12.7

/*
Copyright 2016-2017 Balena

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
exports.osInit = {
  signature: 'internal osinit <image> <type> <config>',
  description: 'do actual init of the device with the preconfigured os image',
  help: 'Don\'t use this command directly! Use `balena os initialize <image>` instead.',
  hidden: true,
  root: true,
  action: function(params, options, done) {
    var Promise, configPromise, helpers, init, manifestPromise;
    Promise = require('bluebird');
    init = require('balena-device-init');
    helpers = require('../utils/helpers');
    configPromise = Promise["try"](function() {
      return JSON.parse(params.config);
    });
    manifestPromise = helpers.getManifest(params.image, params.type);
    return Promise.join(configPromise, manifestPromise, function(config, manifest) {
      return init.initialize(params.image, manifest, config);
    }).then(helpers.osProgressHandler).nodeify(done);
  }
};

exports.scanDevices = {
  signature: 'internal scandevices',
  description: 'scan for local balena-enabled devices and show a picker to choose one',
  help: 'Don\'t use this command directly!',
  hidden: true,
  root: true,
  action: function(params, options, done) {
    var Promise, forms;
    Promise = require('bluebird');
    forms = require('balena-sync').forms;
    return Promise["try"](function() {
      return forms.selectLocalBalenaOsDevice().then(function(hostnameOrIp) {
        return console.error("==> Selected device: " + hostnameOrIp);
      });
    }).nodeify(done);
  }
};

exports.sudo = {
  signature: 'internal sudo <command>',
  description: 'execute arbitrary commands in a privileged subprocess',
  help: 'Don\'t use this command directly!\n\n<command> must be passed as a single argument. That means, you need to make sure\nyou enclose <command> in quotes (eg. balena internal sudo \'ls -alF\') if for\nwhatever reason you invoke the command directly or, typically, pass <command>\nas a single argument to spawn (eg. `spawn(\'balena\', [ \'internal\', \'sudo\', \'ls -alF\' ])`).\n\nFurthermore, this command will naively split <command> on whitespace and directly\nforward the parts as arguments to `sudo`, so be careful.',
  hidden: true,
  action: function(params, options, done) {
    var Promise, os;
    os = require('os');
    Promise = require('bluebird');
    return Promise["try"](function() {
      var cmd, ps, spawn, wait, windosu;
      if (os.platform() === 'win32') {
        windosu = require('windosu');
        return windosu.exec(params.command, {});
      } else {
        spawn = require('child_process').spawn;
        wait = require('rindle').wait;
        cmd = params.command.split(' ');
        ps = spawn('sudo', cmd, {
          stdio: 'inherit',
          env: process.env
        });
        return wait(ps);
      }
    }).nodeify(done);
  }
};
