// balena-cli/build/actions/notes.js
// Generated by CoffeeScript 1.12.7

/*
Copyright 2016-2017 Balena

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
var normalizeUuidProp;

normalizeUuidProp = require('../utils/normalization').normalizeUuidProp;

exports.set = {
  signature: 'note <|note>',
  description: 'set a device note',
  help: 'Use this command to set or update a device note.\n\nIf note command isn\'t passed, the tool attempts to read from `stdin`.\n\nTo view the notes, use $ balena device <uuid>.\n\nExamples:\n\n	$ balena note "My useful note" --device 7cf02a6\n	$ cat note.txt | balena note --device 7cf02a6',
  options: [
    {
      signature: 'device',
      parameter: 'device',
      description: 'device uuid',
      alias: ['d', 'dev'],
      required: 'You have to specify a device'
    }
  ],
  permission: 'user',
  action: function(params, options, done) {
    var Promise, _, balena, exitWithExpectedError;
    normalizeUuidProp(options, 'device');
    Promise = require('bluebird');
    _ = require('lodash');
    balena = require('balena-sdk').fromSharedOptions();
    exitWithExpectedError = require('../utils/patterns').exitWithExpectedError;
    return Promise["try"](function() {
      if (_.isEmpty(params.note)) {
        exitWithExpectedError('Missing note content');
      }
      return balena.models.device.note(options.device, params.note);
    }).nodeify(done);
  }
};
