// balena-cli/build/actions/wizard.js
// Generated by CoffeeScript 1.12.7

/*
Copyright 2016-2017 Balena

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 */
exports.wizard = {
  signature: 'quickstart [name]',
  description: 'getting started with balena',
  help: 'Use this command to run a friendly wizard to get started with balena.\n\nThe wizard will guide you through:\n\n	- Create an application.\n	- Initialise an SDCard with the balena operating system.\n	- Associate an existing project directory with your balena application.\n	- Push your project to your devices.\n\nExamples:\n\n	$ balena quickstart\n	$ balena quickstart MyApp',
  primary: true,
  action: function(params, options, done) {
    var balena, patterns, runCommand;
    balena = require('balena-sdk').fromSharedOptions();
    patterns = require('../utils/patterns');
    runCommand = require('../utils/helpers').runCommand;
    return balena.auth.isLoggedIn().then(function(isLoggedIn) {
      if (isLoggedIn) {
        return;
      }
      console.info('Looks like you\'re not logged in yet!');
      console.info("Let's go through a quick wizard to get you started.\n");
      return runCommand('login');
    }).then(function() {
      if (params.name != null) {
        return;
      }
      return patterns.selectOrCreateApplication().tap(function(applicationName) {
        return balena.models.application.has(applicationName).then(function(hasApplication) {
          if (hasApplication) {
            return applicationName;
          }
          return runCommand("app create " + applicationName);
        });
      }).then(function(applicationName) {
        return params.name = applicationName;
      });
    }).then(function() {
      return runCommand("device init --application " + params.name);
    }).tap(patterns.awaitDevice).then(function(uuid) {
      return runCommand("device " + uuid);
    }).then(function() {
      return balena.models.application.get(params.name);
    }).then(function(application) {
      return console.log("Your device is ready to start pushing some code!\n\nCheck our official documentation for more information:\n\n    http://balena.io/docs/#/pages/introduction/introduction.md\n\nClone an example or go to an existing application directory and run:\n\n    $ git remote add balena " + application.git_repository + "\n    $ git push balena master");
    }).nodeify(done);
  }
};
